File: 11linkedListValues.js

// ITERATIVE
// n = number of nodes
// time:  O(n)
// space: O(n)

const linkedListValues = (head) => {
  // todo
  const values = [];
  let dummy = head;
  
  while (dummy) {
    values.push(dummy.val)
    dummy = dummy.next;
  }
  
  return values;
};


// RECURSIVE
// n = number of nodes
// time:  O(n)
// space: O(n)
const linkedListValues = (head) => {
  // todo
  const values = [];
  _linkedListValues(head, values);
  return values;
};

const _linkedListValues = (head, values) => {
  if (!head) return;
  values.push(head.val);
  _linkedListValues(head.next, values);
};

File: Python/11linked_list_values.py


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(n)

def linked_list_values(head):
  values = []
  current = head
  while current is not None:
    values.append(current.val)
    current = current.next
  return values


# -----------------------------------

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)
def linked_list_values(head):
  values = []
  _linked_list_values(head, values)
  return values


def _linked_list_values(head, values):
  if head is None:
    return
  values.append(head.val)
  _linked_list_values(head.next, values)


# -----------------------------------
# recursive
# what's complexity. This algo creates multiple arrays

def linked_list_values(head):
  if head is None:
    return []

  array = linked_list_values(head.next)
  return [ head.val, *array ]
